### DISCUSSION TITLE
How to wait for page loaded

### DISCUSSION THREAD

**Discussion by mwpowellhtx** [cefsharp/CefSharp#4870] [Q&A]
I am currently watching some events, address changed, and so forth. I wish to respond when I see certain addresses changed, but I am not certain that the page has fully loaded, i.e.

```csharp
if (Browser.IsLoading)
{
    await Browser.WaitForRenderIdleAsync(1000, TimeSpan.FromSeconds(30d));
}
```

Perhaps it has loaded, then I am not sure why this request would make sense at all. Or what is generally recommended to ensure that a page has loaded before being able to probe elements and such. Of course without a crystal ball knowing the page server, architecture, etc, but just in general.

I am getting a time out during this request for whatever reason. So it seems the page _"is loading"_ so cannot exactly depend on some of the elements being there yet. But for it to just time out, I'm not sure what the parameters are telling us, the idle time? Justaposed with the timeout?

I have some vague notions maybe what I could tweak, but I'm afraid I would be shooting blinding in the dark. Any pointers would be appreciated.

Thanks!

**Comment by mwpowellhtx** [comment]
Or perhaps would it be waiting for navigation, when `Browser.IsLoading`...

```csharp
await Browser.WaitForNavigationAsync();
```

Because obviously if we just assume the new address is loaded and start jamming through elements, we are likely to encounter unexpected errors.

**Comment by mwpowellhtx** [comment]
What we are doing specifically. The page has already loaded, but the elements we are interested are being AJAX loaded, as far as we can determine, after the fact.

```csharp
// Depends upon the bot status <div id="..."/> and back out to the associated server breadcrumbs.
var divBotStatus = await context.QuerySelectorAsync<HtmlDivElement>(S.BotStatus);

// TODO: do not necessarily want to get caught in a tight loop here
// TODO: but do not know what else to do right now
int badgeAttempts = 0;
HtmlSpanElement spanBadge = null;

// TODO: I think we will legit "not" see this span because it is AJAX filled in after the page has "loaded"
while (spanBadge is null)
{
    //// TODO: may may not need to refresh the bot status div...
   //divBotStatus = await context.QuerySelectorAsync<HtmlDivElement>(S.BotStatus)
    ++badgeAttempts;
    spanBadge = await divBotStatus.QuerySelectorAsync<HtmlSpanElement>(S.SpanBadge);
}

var spanClassList = await spanBadge.GetClassListAsync();
```

`spanBadge` does eventually load, but it is `null` to begin with. In one debug session attempts were several hundred, `271` to be precise.

Is there a better way to wait for the badge to arrive, `async` `await` that message, etc?

And whether successful or warning or what not, we need that status for subsequent parts of the application to be available, let alone useful.

**Reply by mwpowellhtx** [reply]
Turns out I think we can wait for the badge and the attempts is just `1`. Will probably work. For now; caveat we do not own the target web site, but we do subscribe there.

```csharp
var spanBadgeEH = await context.WaitForSelectorAsync(S.SpanBadge);

int badgeAttempts = 0;
HtmlSpanElement spanBadge = null;

while (spanBadgeEH is not null && spanBadge is null)
{
    ++badgeAttempts;
    spanBadge = await divBotStatus.QuerySelectorAsync<HtmlSpanElement>(S.SpanBadge);
}
```

**Comment by amaitland** [comment]
> but I am not certain that the page has fully loaded,

That's a tough one. Fully loaded is often very different depending on the page.

There are a number of different strategies, which to use depends on the use case

## Basic (built in options).

`WaitForRenderIdleAsync` will always timeout on a page that has for an example a looping animation, rendering never stops.

`WaitForNavigationAsync` will wait until the navigation has completed, the `DOM` should be loaded by this point (`DOMContentLoaded`). For pages that are generated dynamically in javascript, this is going to be triggered way too early.

`WaitForSelectorAsync` (there's a basic version included directly in the library) for waiting for a `DOM` element to be present, this is the most common approach that I would use.
https://github.com/cefsharp/CefSharp/blob/19bb62359afdcd4df1ea99a99b7b0b86925c5b98/CefSharp/WebBrowserExtensions.cs#L624

## Advanced (CefSharp.Dom)

- You can disable CEF's request interceptor and rely on the CDP based one in `CefSharp.Dom`. There are some more advanced options (which were ported from puppeteer-sharp).

- WaitForNetworkIdleAsync
- WaitForRequestAsync
- WaitForResponseAsync
- WaitForXPathAsync
- WaitForSelectorAsync
- WaitForFunctionAsync
https://github.com/cefsharp/CefSharp.Dom/blob/dc7f2928e53afcf1061e3026439362ef269b7a04/lib/PuppeteerSharp/DevToolsContext.cs#L1576C27-L1576C50

**Reply by mwpowellhtx** [reply]
Thanks for the response and various `WaitForXYZ` bits. Indeed, yes, very DOM specific. I am finding that although the page may be loaded, there are certain elements that are injected AJAX style after the fact. I think I have identified what those are and am waiting for those specifically, seems to be working well enough at the appropriate strategic moments, for now. Of course major caveat, I am not the owner of the source Url in question, so that being said. Rinse and repeat taking into consideration DOM specifics.
